### ISSUE TITLE
CA1812 fires on internal value convertes in WPF projects

### ISSUE THREAD

**Issue by the-black-wolf** [dotnet/roslyn-analyzers#3288]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.8 (Latest)

#### Diagnostic ID

Example: [CA1812](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1812?view=vs-2019)

#### Repro steps

1. Create wpf project in .net core
2. Add custom value converter (e.g.
```csharp
internal sealed class DeleteButtonVisibilityConverter : IMultiValueConverter
```
3. Add value converter on the WPF page
```xml
<DataGrid.Resources>
    <conv:DeleteButtonVisibilityConverter x:Key="DeleteButtonVisibilityConverter"/>
</DataGrid.Resources>

...
<MultiBinding Converter="{StaticResource DeleteButtonVisibilityConverter}" Mode="OneWay">
```

#### Expected behavior
No warning for CA181, because object is instantiated in WPF page

#### Actual behavior
CA1812 fires, seems to be completely oblivious to XAML. Supressing this issue is a workaround (I guess its a workaround for any CA issue), but then it hides actual problems in WPF project. 
The same problem appears on various DI containers.

**Comment by Evangelink** [comment]
Looking at the rule implementation I assume that the analyzer is triggered on the intermediate C# file generated from the xaml and so that the call shall be detected.

**Comment by billybednar** [comment]
Does it go away if you open the file where the type is declared? I've encountered this in a .NET Framework WPF project, but it only occurs after a clean build. The diagnostic magically disappears after opening the file where the type is declared.

**Comment by cjmurph** [comment]
> Does it go away if you open the file where the type is declared? I've encountered this in a .NET Framework WPF project, but it only occurs after a clean build. The diagnostic magically disappears after opening the file where the type is declared.

No, it doesn't go away for me.

CodeLens shows zero references on the class too.

**Comment by Evangelink** [comment]
@the-black-wolf @cjmurph I can reproduce the behavior.

@mavasani The analyzer is configured to [analyze generated code](https://github.com/dotnet/roslyn-analyzers/blob/master/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs#L43) but it seems that the converter doesn't appear in the `***.g.cs` nor the `***.g.i.cs`. There data is either contained in some other document I am missing or the analyzer would also need references to xaml files which seems like a bigger task. Any idea?

**Comment by billybednar** [comment]
Disregard my earlier comment about the diagnostic disappearing. That's an unrelated issue (see roslyn#19693 and roslyn#26817), not the analyzer detecting references in XAML only sometimes.

The reason the analyzer isn't detecting references in XAML is because they don't result in any code generation. Instead, XAML is converted to a binary format and embedded as a resource. WPF reads that at runtime and use the information to create the object graph using reflection.

**Comment by Evangelink** [comment]
@mavasani Is there anything that can be done from the point of view of this repo? Reading the various threads linked it seems more of a global issue than a roslyn-analyzer one.

**Comment by mavasani** [comment]
This looks related to https://github.com/dotnet/roslyn/issues/38192 - likely the code behind file for xaml which has the converter instantiation is not generated? Any case, I don't think there is anything that the analyzer can specifically do here.
